WalletWasabi/Services/SingleInstanceChecker.cs

```diff
@@ -0,0 +1,58 @@
+using NBitcoin;
+using Nito.AsyncEx;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace WalletWasabi.Services
+{
+	public class SingleInstanceChecker : IDisposable
+	{
+		private const string MutexString = "WalletWasabiSingleInstance";
+		private bool _disposedValue;
+
+		public SingleInstanceChecker(Network network)
+		{
+			Network = network;
+		}
+
+		private IDisposable SingleApplicationLockHolder { get; set; }
+		private Network Network { get; }
+
+		public async Task CheckAsync()
+		{
+			using CancellationTokenSource cts = new CancellationTokenSource(TimeSpan.Zero);
+
+			// The disposal of this mutex handled by AsyncMutex.WaitForAllMutexToCloseAsync().
+			var mutex = new AsyncMutex($"{MutexString}-{Network}");
+			try
+			{
+				SingleApplicationLockHolder = await mutex.LockAsync(cts.Token).ConfigureAwait(false);
+			}
+			catch (IOException)
+			{
+				throw new InvalidOperationException($"Wasabi is already running on {Network}!");
```

Comment:
Why do we specify the network here?
Can we run two instances of Wasabi on different networks?

Otherwise this tACK.

Reply:
I think so, they still will share the Logs.txt file but that should be all.

Reply:
But what if you open the same wallet on different networks? This can mess up things.
The point I am trying to make is we should simply say `Wasabi is already running`, but I think this is not very important because this message will not be displayed to the end user.

Reply:
This was something that @nopara73  asked and we decided to do. I think it is an edge case as it is hard to run wasabi on different networks you have to hack the config file.

Reply:
It's nice to propagate the exception when it is possible:

```csharp
catch (IOException e)
{
	throw new InvalidOperationException($"Wasabi is already running on {Network}!", e);
}   
```